common/lib/api-client/middleware/got-request.js

```diff
@@ -39,6 +39,10 @@ function requestMiddleware({
           http: new HttpAgent(),
           https: new HttpsAgent(),
         },
+        retry: {
+          limit: 1,
+          statusCodes: [502, 504],
+        },
```

Comment:
Is there a danger of retrying a post request here? Maybe @martyn-w best placed to advise, but the API throws those errors sometimes from prison API, and the call to that sometimes happens after the API has saved some data into the database. I do not believe API requests are idempotent, but I am not 100% sure. If I am right, then retrying could leave us with duplicate data. Maybe?

Reply:
That's a good question. 

However, the calls to build the dashboards and incoming/outgoing pages are POST requests also so if we remove `POST`s from the retries we wouldn't retry those which is an error we often see these type of failures.

Reply:
Please note - this is another reason why those calls should not be POSTs. I would rather *not* retry these than risk retrying a POST we don't want to carry out twice

Reply:
I agree - the only calls which can be safely retried are `GET` requests, plus the following three `POST` requests:
* `/api/allocations/filtered`
* `/api/moves/filtered`
* `/api/moves/csv`

PATCH / DELETE / PUT and any other POST should not be retried otherwise there is a risk of duplicate / corrupt data.

The reason for the three POST routes for getting data is to do with the length limitation on URL query parameters: I think the front end needs to send a large number of location IDs for getting filtered moves - the length of this list exceeded the maximum URL length.

Reply:
>The reason for the three POST routes for getting data is to do with the length limitation on URL query parameters: I think the front end needs to send a large number of location IDs for getting filtered moves - the length of this list exceeded the maximum URL length.

@martyn-w that's correct, we had to move to a `POST` to allow us to not have to batch up these requests and reduce the calls needed to build the dashboards.

@Nimphal @martyn-w @martin-ballhatchet-moj Shall we have a quick 5 mins to discuss this today? Be good to understand the details of what exactly could happen on retrying some of the requests with data and if there's way to prevent duplication within the API. If the request failed downstream and we bubble the error back to the user on the frontend, we currently tell them to "retry again" anyway, so whether its this layer auto-retrying or the user trying again, if there is a risk of duplication it will occur in both instances.

Reply:
"Be good to understand the details of what exactly could happen on retrying some of the requests with data" - that's a significant piece of work, for very little benefit. Based upon the general consensus, we can't retry POSTs. In case it helps - I think we were only talking about retrying 5xx errors against HMPPS Auth initially - based upon the advice of the HMPPS Auth team. If there's benefit to retrying 5xx for the API, I'm OK with it being done for GETs, but nothing else.

Reply:
>that's a significant piece of work, for very little benefit

In the instance of a failed `POST` or `PATCH` with a `502` the user is still able to manually retry by refreshing the page, so if we're worried about duplicate data being created, it can still occur whether its a retry from our request library or the user.

The issue with only retrying `GET`s at this stage is I don't think it will resolve the [Bad Gateway errors](https://sentry.io/organizations/ministryofjustice/issues/2395154994/events/?project=2014813&query=Bad+Gateway&statsPeriod=14d) we've seen in the frontend coming from the API. That majority have come from either the dashboard or incoming/outgoing dashboards which fail on the `POST` filters.

The frequency does seem to be low though, so if we're happy to just accept them occurring and ignoring them we can also do that.